python/mpt/src/mpt/trie_diff.py

```diff
@@ -0,0 +1,201 @@
+                # Look for diffs in the left sub-tree
+                l_node: ExtensionNode = cast(ExtensionNode, l_node)
+                self._compute_diff(l_node.subnode, None, path + l_node.key_segment)
+
+            case "ExtensionNode", "LeafNode":
+                raise ValueError("ExtensionNode -> LeafNode should not happen if comparing two valid trie roots")
+
+            case "ExtensionNode", "ExtensionNode":
+                l_node: ExtensionNode = cast(ExtensionNode, l_node)
+                r_node: ExtensionNode = cast(ExtensionNode, r_node)
+
+                # Equal keys -> Look for diffs in childrens
+                if l_node.key_segment == r_node.key_segment:
+                    self._compute_diff(l_node.subnode, r_node.subnode, path + l_node.key_segment)
+
+                # Right is prefix of left -> Look for diffs in left sub-tree
+                elif l_node.key_segment.startswith(r_node.key_segment):
+                    # Remove the prefix from the extension key segment
+                    l_node.key_segment = l_node.key_segment[len(r_node.key_segment):]
+                    self._compute_diff(l_node.subnode, None, path + l_node.key_segment)
+
+                # Left is prefix of right -> Look for diffs in right sub-tree
+                elif r_node.key_segment.startswith(l_node.key_segment):
+                    # Remove the prefix from the extension key segment
+                    r_node.key_segment = r_node.key_segment[len(l_node.key_segment):]
+                    self._compute_diff(None, r_node.subnode, path + r_node.key_segment)
+
+                # Both are different -> Look for diffs in both sub-trees
+                else:
+                    self._compute_diff(l_node.subnode, r_node.subnode, path + l_node.key_segment)
+
+            case "ExtensionNode", "BranchNode":
+                l_node: ExtensionNode = cast(ExtensionNode, l_node)
+                r_node: BranchNode = cast(BranchNode, r_node)
+                # Match on the corresponding nibble of the extension key segment
+                for i in range(0, 16):
+                    nibble = i.to_bytes(1, "big")
+                    if l_node.key_segment[0] == nibble:
+                        # Remove the nibble from the extension key segment
+                        l_node.key_segment = l_node.key_segment[1:]
+                        self._compute_diff(l_node.subnode, r_node.subnodes[i], path + nibble)
+                    else:
+                        # Look for diffs in other branches
+                        self._compute_diff(l_node.subnode, r_node.subnodes[i], path + nibble)
+
+            case "BranchNode", "None":
+                # Look for diffs in all branches of the left sub-tree
+                l_node: BranchNode = cast(BranchNode, l_node)
+                for i in range(0, 16):
+                    self._compute_diff(l_node.subnodes[i], None, path + i.to_bytes(1, "big"))
+
+            case "BranchNode", "LeafNode":
```

Comment:
why not?

branch with two branches, both branches leading to a leaf
if you delete one branch, then you have to perform a branch node reduction into a single leaf